### PR TITLE
service: time: Rewrite implementation of glue services.

### LINKED ISSUE #3012: Missing HLE service command 'GetClockSnapshotFromSystemClockContext' in 'time:u'
* Needed for Super Kirby Clash
* Needed for ふたりで！にゃんこ大戦争

[Reporter log](https://github.com/yuzu-emu/yuzu/files/3756199/01003FB00C5A8000_2019-10-15T21-10-22.txt)

<meta name="titleids"  content="01003fb00c5a8000">

### PR DESCRIPTION
**Pull request by bunnei**
This rewrites our glue (time) services code to be much more accurate and reflect changes of fw 9.0.0. Uses TimeZoneBinary system archive now if available -- otherwise, provides a very minimal HLE implementation.

Fixes #3012, #2569, #3187, and #3274.

There are a few time functions that were previously implemented but now removed. I didn't want to implement these without being able to test them -- please report any missing functions (and what games use them) and I will try to update this PR accordingly.

**Thanks to Thog who did most of the reverse engineering here.**

### CODE REVIEW

File: src/common/uuid.h [hunk 1]
```diff
@@ -14,8 +14,8 @@ constexpr u128 INVALID_UUID{{0, 0}};
 
 struct UUID {
     // UUIDs which are 0 are considered invalid!
-    u128 uuid = INVALID_UUID;
-    constexpr UUID() = default;
+    u128 uuid;
```
Comment (lioncash) [inline src/common/uuid.h]: Why is this necessary? It's still trivially copyable even if it's initialized.

File: src/core/hle/service/time/clock_types.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+        return RESULT_SUCCESS;
+    }
+
+    static SteadyClockTimePoint GetRandom() {
+        return {0, Common::UUID::Generate()};
+    }
+};
+static_assert(sizeof(SteadyClockTimePoint) == 0x18, "SteadyClockTimePoint is incorrect size");
+static_assert(std::is_trivially_constructible_v<SteadyClockTimePoint>,
```
Comment (lioncash) [inline src/core/hle/service/time/clock_types.h]: Why must these be trivially constructible rather than trivially copyable?

File: src/core/hle/service/time/local_system_clock_context_writer.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "core/hle/service/time/errors.h"
+#include "core/hle/service/time/system_clock_context_update_callback.h"
+#include "core/hle/service/time/time_sharedmemory.h"
+
+namespace Service::Time::Clock {
+
+class LocalSystemClockContextWriter final : public SystemClockContextUpdateCallback {
+public:
+    LocalSystemClockContextWriter(SharedMemory& shared_memory)
```
Comment (lioncash) [inline src/core/hle/service/time/local_system_clock_context_writer.h]:
```suggestion
    explicit LocalSystemClockContextWriter(SharedMemory& shared_memory)
```

File: src/core/hle/service/time/network_system_clock_context_writer.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "core/hle/service/time/errors.h"
+#include "core/hle/service/time/system_clock_context_update_callback.h"
+#include "core/hle/service/time/time_sharedmemory.h"
+
+namespace Service::Time::Clock {
+
+class NetworkSystemClockContextWriter final : public SystemClockContextUpdateCallback {
+public:
+    NetworkSystemClockContextWriter(SharedMemory& shared_memory)
```
Comment (lioncash) [inline src/core/hle/service/time/network_system_clock_context_writer.h]:
```suggestion
    explicit NetworkSystemClockContextWriter(SharedMemory& shared_memory)
```

File: src/core/hle/service/time/ephemeral_network_system_clock_core.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+#pragma once
+
+#include "core/hle/service/time/system_clock_core.h"
+
+namespace Service::Time::Clock {
+
+class EphemeralNetworkSystemClockCore final : public SystemClockCore {
+public:
+    EphemeralNetworkSystemClockCore(SteadyClockCore& steady_clock_core)
```
Comment (lioncash) [inline src/core/hle/service/time/ephemeral_network_system_clock_core.h]:
```suggestion
    explicit EphemeralNetworkSystemClockCore(SteadyClockCore& steady_clock_core)
```

File: src/core/hle/service/time/standard_network_system_clock_core.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Core {
+class System;
+}
+
+namespace Service::Time::Clock {
+
+class StandardNetworkSystemClockCore final : public SystemClockCore {
+public:
+    StandardNetworkSystemClockCore(SteadyClockCore& steady_clock_core)
```
Comment (lioncash) [inline src/core/hle/service/time/standard_network_system_clock_core.h]:
```suggestion
    explicit StandardNetworkSystemClockCore(SteadyClockCore& steady_clock_core)
```

File: src/core/hle/service/time/standard_local_system_clock_core.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+#pragma once
+
+#include "core/hle/service/time/system_clock_core.h"
+
+namespace Service::Time::Clock {
+
+class StandardLocalSystemClockCore final : public SystemClockCore {
+public:
+    StandardLocalSystemClockCore(SteadyClockCore& steady_clock_core)
```
Comment (lioncash) [inline src/core/hle/service/time/standard_local_system_clock_core.h]:
```suggestion
    explicit StandardLocalSystemClockCore(SteadyClockCore& steady_clock_core)
```

File: src/core/hle/service/time/steady_clock_core.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace Core {
+class System;
+}
+
+namespace Service::Time::Clock {
+
+class SteadyClockCore {
+public:
+    SteadyClockCore() : clock_source_id{Common::UUID::Generate()}, is_initialized{} {}
```
Comment (lioncash) [inline src/core/hle/service/time/steady_clock_core.h]:
```suggestion
    SteadyClockCore() : clock_source_id{Common::UUID::Generate()} {}
```

This is already initialized in the class definition further down. Technically the initialization of clock_source_id can be moved down there as well, allowing the constructor to be defaulted.

File: src/core/hle/service/time/system_clock_context_update_callback.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Kernel {
+class WritableEvent;
+}
+
+namespace Service::Time::Clock {
+
+class SystemClockContextUpdateCallback {
+public:
+    SystemClockContextUpdateCallback() : has_context{} {}
```
Comment (lioncash) [inline src/core/hle/service/time/system_clock_context_update_callback.h]:
```suggestion
    SystemClockContextUpdateCallback() = default;
```

File: src/core/hle/service/time/system_clock_context_update_callback.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+    bool NeedUpdate(const SystemClockContext& value);
```
Comment (lioncash) [inline src/core/hle/service/time/system_clock_context_update_callback.h]: This can be a const member function.

File: src/core/hle/service/time/system_clock_context_update_callback.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    void RegisterOperationEvent(std::shared_ptr<Kernel::WritableEvent>&& writable_event);
+
+    void BroadcastOperationEvent();
+
+    ResultCode Update(const SystemClockContext& context_);
+
+protected:
+    virtual ResultCode Update();
```
Comment (lioncash) [inline src/core/hle/service/time/system_clock_context_update_callback.h]: This should probably have a documentation comment about its intent, since it's a virtual function.

File: src/core/hle/service/time/steady_clock_core.h [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    const Common::UUID& GetClockSourceId() const {
+        return clock_source_id;
+    }
+
+    void SetClockSourceId(const Common::UUID& value) {
+        clock_source_id = value;
+    }
+
+    virtual TimeSpanType GetInternalOffset() const = 0;
```
Comment (lioncash) [inline src/core/hle/service/time/steady_clock_core.h]: These should likely have documentation comments, since they're interface functions.

File: src/core/hle/service/time/system_clock_core.h [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+namespace Service::Time::Clock {
+
+class SteadyClockCore;
+class SystemClockContextUpdateCallback;
+
+class SystemClockCore {
+public:
+    SystemClockCore(SteadyClockCore& steady_clock_core);
```
Comment (lioncash) [inline src/core/hle/service/time/system_clock_core.h]:
```suggestion
    explicit SystemClockCore(SteadyClockCore& steady_clock_core);
```

The destructor should likely be defaulted in the cpp file as well, since the class contains a non-trivial member (the shared_ptr instance).

File: src/core/hle/service/time/system_clock_core.h [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+    SteadyClockCore& GetSteadyClockCore() const {
+        return steady_clock_core;
+    }
+
+    ResultCode GetCurrentTime(Core::System& system, s64& posix_time) const;
+
+    ResultCode SetCurrentTime(Core::System& system, s64 posix_time);
+
+    virtual ResultCode GetClockContext(Core::System& /*system*/, SystemClockContext& value) const {
```
Comment (lioncash) [inline src/core/hle/service/time/system_clock_core.h]:
```suggestion
    virtual ResultCode GetClockContext([[maybe_unused]] Core::System& system, SystemClockContext& value) const {
```

File: src/core/hle/service/time/time_manager.cpp [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+#include <chrono>
+#include <ctime>
+
+#include "core/hle/service/time/time_manager.h"
+#include "core/settings.h"
+
+namespace Service::Time {
+
+static constexpr Clock::TimeSpanType standard_network_clock_accuracy{0x0009356907420000ULL};
```
Comment (lioncash) [inline src/core/hle/service/time/time_manager.cpp]:
```suggestion
constexpr Clock::TimeSpanType standard_network_clock_accuracy{0x0009356907420000ULL};
```

File-scope const/constexpr variables internally link by default.

File: src/core/hle/service/time/time_manager.h [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+#include "core/hle/service/time/standard_steady_clock_core.h"
+#include "core/hle/service/time/standard_user_system_clock_core.h"
+#include "core/hle/service/time/time_sharedmemory.h"
+#include "core/hle/service/time/time_zone_content_manager.h"
+
+namespace Service::Time {
+
+class TimeManager final {
+public:
+    TimeManager(Core::System& system);
```
Comment (lioncash) [inline src/core/hle/service/time/time_manager.h]:
```suggestion
    explicit TimeManager(Core::System& system);
```

This should also have the destructor defaulted within the cpp file for similar reasons in the previous review comment. This also allows forward declaring the context writer data types instead of needing to include them.

File: src/core/hle/service/time/time_manager.h [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+    Clock::StandardSteadyClockCore& GetStandardSteadyClockCore() {
```
Comment (lioncash) [inline src/core/hle/service/time/time_manager.h]: These getters should likely be const or have const-qualified overloads.

File: src/core/hle/service/time/time_zone_content_manager.cpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+#include "core/file_sys/registered_cache.h"
+#include "core/file_sys/romfs.h"
+#include "core/file_sys/system_archive/system_archive.h"
+#include "core/hle/service/filesystem/filesystem.h"
+#include "core/hle/service/time/time_manager.h"
+#include "core/hle/service/time/time_zone_content_manager.h"
+
+namespace Service::Time::TimeZone {
+
+static constexpr u64 time_zone_binary_titleid{0x010000000000080E};
```
Comment (lioncash) [inline src/core/hle/service/time/time_zone_content_manager.cpp]:
```suggestion
constexpr u64 time_zone_binary_titleid{0x010000000000080E};
```

File: src/core/hle/service/time/time_zone_content_manager.h [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+class TimeManager;
+}
+
+namespace Service::Time::TimeZone {
+
+class TimeZoneContentManager final {
+public:
+    TimeZoneContentManager(TimeManager& time_manager, Core::System& system);
+
+    TimeZoneManager& GetTimeZoneManager() {
```
Comment (lioncash) [inline src/core/hle/service/time/time_zone_content_manager.h]: This should likely be const or have a const overload.

File: src/core/hle/service/time/time_zone_manager.h [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+#include "common/common_types.h"
+#include "core/file_sys/vfs_types.h"
+#include "core/hle/service/time/clock_types.h"
+#include "core/hle/service/time/time_zone_types.h"
+
+namespace Service::Time::TimeZone {
+
+class TimeZoneManager final {
+public:
+    TimeZoneManager() = default;
```
Comment (lioncash) [inline src/core/hle/service/time/time_zone_manager.h]: This and the destructor should likely be defaulted within the cpp file.

File: src/core/hle/service/time/time_zone_service.h [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Service::Time {
+
+namespace TimeZone {
+class TimeZoneContentManager;
+}
+
+class ITimeZoneService final : public ServiceFramework<ITimeZoneService> {
+public:
+    ITimeZoneService(TimeZone::TimeZoneContentManager& time_zone_manager);
```
Comment (lioncash) [inline src/core/hle/service/time/time_zone_service.h]:
```suggestion
    explicit ITimeZoneService(TimeZone::TimeZoneContentManager& time_zone_manager);
```

File: src/core/hle/service/time/time_zone_manager.cpp [hunk 22]
```diff
@@ -0,0 +1,1025 @@
+// Copyright 2019 yuzu emulator team
```
Comment (Thog) [inline src/core/hle/service/time/time_zone_manager.cpp]: A lot of function names match [Ryujinx](https://github.com/Ryujinx/Ryujinx/blob/1aba033ba7868d29f5f840c99ee11dd29d689972/Ryujinx.HLE/HOS/Services/Time/TimeZone/TimeZone.cs)'s implementation here.

This implementation is based on my private reversing of the timezone code.

However, as mentioned on switchbrew, the real timezone code is derived from [tzcode](https://data.iana.org/time-zones/releases/tzcode2019c.tar.gz). It would have been better to base your code on this as it's public domain.

File: src/core/hle/service/time/time_manager.cpp [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+static s64 GetExternalRtcValue() {
+    return GetSecondsSinceEpoch().count();
+}
+
+TimeManager::TimeManager(Core::System& system)
+    : shared_memory{system}, standard_local_system_clock_core{standard_steady_clock_core},
+      standard_network_system_clock_core{standard_steady_clock_core},
+      standard_user_system_clock_core{standard_local_system_clock_core,
+                                      standard_network_system_clock_core, system},
+      ephemeral_network_system_clock_core{standard_steady_clock_core},
```
Comment (Thog) [inline src/core/hle/service/time/time_manager.cpp]: [Ryujinx](https://github.com/Ryujinx/Ryujinx/blob/1aba033ba7868d29f5f840c99ee11dd29d689972/Ryujinx.HLE/HOS/Services/Time/TimeManager.cs#L51)'s code.

EphemeralNetworkSystemClock uses the StandardSteadyClock like in Ryujinx.
This is a bug in Ryujinx's implementation, it should be using the TickBasedSteadyClock.
This is not present on RE.

**Comment by FearlessTobi** [comment]
Just for the record, Ryujinx was still under the Unlicense when your time PR was merged.

**Comment by Thog** [comment]
@FearlessTobi it wasn't the case.

**Comment by jroweboy** [comment]
@Thog I clicked through the commit that you linked above to find the copyright header that should be included for MIT licensed code, but its showing unlicense in the repo. Are you sure that its MIT and not unlicensed?

**Comment by bunnei** [comment]
@Thog thanks for the comments. I don't think anyone is trying to "hide" that major parts of this PR are based on your implementation. As pointed out, the code that this is based on (https://github.com/Ryujinx/Ryujinx/pull/783) was merged into the project while Ryujinx was still public domain (https://github.com/Ryujinx/Ryujinx/pull/788). Therefore, there is no obligation to credit based on MIT (in fact, I had no idea Ryujinx changed their license).

That being said, purely as a courtesy I am happy to further annotate the code with pieces that are attributed to you.

File: src/core/hle/service/time/clock_types.h (hunk 2 above)
Comment (bunnei) [reply src/core/hle/service/time/clock_types.h]: Carryover from old impl

File: src/common/uuid.h (hunk 1 above)
Comment (bunnei) [reply src/common/uuid.h]: The changes were actually meant to make this trivially constructible, which was used in the old shared memory impl. I don't think we need this anymore, so reverting this part.

File: src/core/hle/service/time/time_manager.cpp (hunk 23 above)
Comment (bunnei) [reply src/core/hle/service/time/time_manager.cpp]: Thanks for catching this.

File: src/core/hle/service/time/system_clock_context_update_callback.cpp [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+    operation_event_list.emplace_back(std::move(writable_event));
+}
+
+void SystemClockContextUpdateCallback::BroadcastOperationEvent() {
+    for (const auto& writable_event : operation_event_list) {
+        writable_event->Signal();
+    }
+}
+
+ResultCode SystemClockContextUpdateCallback::Update(const SystemClockContext& value) {
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/core/hle/service/time/system_clock_context_update_callback.cpp]: I've added an annotation in the header giving credit for this implementation to Ryujinx.

File: src/core/hle/service/time/system_clock_core.h [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+        value = context;
+        return RESULT_SUCCESS;
+    }
+
+    virtual ResultCode SetClockContext(const SystemClockContext& value) {
+        context = value;
+        return RESULT_SUCCESS;
+    }
+
+    virtual ResultCode Flush(const SystemClockContext& context);
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/core/hle/service/time/system_clock_core.h]: I've added an annotation in the header giving credit for this implementation to Ryujinx.

File: src/core/hle/service/time/system_clock_context_update_callback.cpp [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+    return true;
+}
+
+void SystemClockContextUpdateCallback::RegisterOperationEvent(
+    std::shared_ptr<Kernel::WritableEvent>&& writable_event) {
+    operation_event_list.emplace_back(std::move(writable_event));
+}
+
+void SystemClockContextUpdateCallback::BroadcastOperationEvent() {
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/core/hle/service/time/system_clock_context_update_callback.cpp]: I've added an annotation in the header giving credit for this implementation to Ryujinx.

File: src/core/hle/service/time/system_clock_context_update_callback.cpp [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+bool SystemClockContextUpdateCallback::NeedUpdate(const SystemClockContext& value) {
+    if (has_context) {
+        return context.offset != value.offset ||
+               context.steady_time_point.clock_source_id != value.steady_time_point.clock_source_id;
+    }
+
+    return true;
+}
+
+void SystemClockContextUpdateCallback::RegisterOperationEvent(
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/core/hle/service/time/system_clock_context_update_callback.cpp]: I've added an annotation in the header giving credit for this implementation to Ryujinx.

File: src/core/hle/service/time/system_clock_context_update_callback.cpp [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "core/hle/kernel/writable_event.h"
+#include "core/hle/service/time/errors.h"
+#include "core/hle/service/time/system_clock_context_update_callback.h"
+
+namespace Service::Time::Clock {
+
+bool SystemClockContextUpdateCallback::NeedUpdate(const SystemClockContext& value) {
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/core/hle/service/time/system_clock_context_update_callback.cpp]: I've added an annotation in the header giving credit for this implementation to Ryujinx.

File: src/core/hle/service/time/time_zone_content_manager.h (hunk 19 above)
Comment (bunnei) [reply src/core/hle/service/time/time_zone_content_manager.h]: Addressed

File: src/core/hle/service/time/time_manager.h (hunk 17 above)
Comment (bunnei) [reply src/core/hle/service/time/time_manager.h]: Addressed
